torch/distributed/_shard/sharding_spec/chunk_sharding_spec.py

```diff
@@ -93,7 +93,7 @@ def build_metadata(self,
             chunked_dim_size = get_chunked_dim_size(sharding_dim_size, split_size, idx)
             shard_size = list(tensor_sizes)
             current_offsets = [0] * tensor_num_dim
-            current_offsets[self.dim] = split_size * idx  # type: ignore[index]
+            current_offsets[self.dim] = min(sharding_dim_size, split_size * idx)  # type: ignore[index]
```

Comment:
`ChunkShardingSpec` says:
```
... it represents sharding a Tensor along a single dimension into equal chunks (similar to :meth:`torch.chunk`).
```

Not sure changing this would break systems (e.g. Model Store?) that relies on the property.

Any chance `EnumerableShardingSpec` would work for this case?

Reply:
`ChunkShardingSpec`  was deduced in `ShardedTensor.init_from_local_shards_and_global_metadata`.
If `ChunkShardingSpec` was not supposed to be used with uneven shards(but sizes are already calculated uneven), we need to change the deduction of Spec in `_infer_sharding_spec_from_shards_metadata` https://github.com/pytorch/pytorch/blob/main/torch/distributed/_shard/sharding_spec/api.py#L169

Reply:
A that's useful info. Do you think adding a condition for checking uniform shard size here would work? https://github.com/pytorch/pytorch/blob/main/torch/distributed/_shard/sharding_spec/api.py#L212

My memory is a bit rusty on this, does our infra support `EnumerableShardingSpec`?

Reply:
I do not know about EnumerableShardingSpec support.
But empty and uneven shards are valid for Torchrec usecase and used.
I will add offsets comparison to spec deduction.

Reply:
@yifuwang I don't know too much historical context on the original design, but it seems a little bit restrictive to allow ChunkShardingSpec to be uniform. I just checked for FSDP this issue doesn't appear, because the metatdata (size, offsets ...) is manually being created here. https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/_shard_utils.py#L26

For the uneven-sharded case, it seems ShardedTensor is ok with taking the uneven offsets and sizes and generating ChunkShardingSpec behind the scenes. 
https://github.com/pytorch/pytorch/blob/main/torch/distributed/_shard/sharded_tensor/api.py#L721

Therefore, I think this fix should be fine, but we can wait for the sandcastle tests to confirm that it doesn't break anything on modelstore.